### PR TITLE
E2E deploy test: rewrite in go tests

### PR DESCRIPTION
**Pull request by rinx**
### Description:

- E2E deploy tests re-written in golang.
- E2E deploy tests will be run by labeling PRs with `actions/e2e-deploy`.

### Related Issue:

Nothing.

### How Has This Been Tested?:

Nothing.

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [X] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [X] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by rinx** [comment]
:memo: after #740 merged into master, this PR will be revised to use v1 APIs for E2E CRUD tests.

**Comment by rinx** [comment]
/label actions/e2e-deploy
fixed the internal/db/rdb/mysql codes just for workaround. it should be revised later.

**Comment by rinx** [comment]
as shown in the FOSSA result above, the newly added dependencies doesn't affect the license issues.

File: .github/workflows/e2e-deploy.yml [hunk 1]
```diff
@@ -93,58 +125,307 @@ jobs:
...
+          kubectl cluster-info
+      - uses: azure/setup-helm@v1
+        with:
+          version: ${{ steps.version.outputs.helm }}
+      - name: Helm version
+        run: |
+          helm version
+      - name: deploy vald
+        run: |
+          make k8s/external/scylla/deploy
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:250]: how about change members to 1. I think which accelarete the CI speed
https://github.com/vdaas/vald/blob/master/k8s/external/scylla/scyllacluster.yaml#L98

File: .github/workflows/e2e-deploy.yml [hunk 2]
```diff
@@ -93,58 +125,307 @@ jobs:
...
+          kubectl cluster-info
+      - uses: azure/setup-helm@v1
+        with:
+          version: ${{ steps.version.outputs.helm }}
+      - name: Helm version
+        run: |
+          helm version
+      - name: deploy vald
+        run: |
+          kubectl apply -f k8s/external/minio
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:348]: I used to thougt using minio for S3 backend. seems very useful could you please add make file command like scylla

File: .github/workflows/e2e-deploy.yml [hunk 3]
```diff
@@ -93,58 +125,307 @@ jobs:
...
+          jq_query='.items[] | select( ([ .status.conditions[] | select( .type == "Ready" and .status == "True" ) ] | length ) == 1) | .metadata.namespace + "/" + .metadata.name'
+          until [ $(kubectl get pod --selector=app=minio -o json | jq -r "$jq_query" | wc -l) -ge 1 ]
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:350]: I think it would be better to wait in make command that i commented

File: go.mod [hunk 4]
```diff
@@ -29,6 +29,7 @@ replace (
 	k8s.io/api => k8s.io/api v0.20.1
 	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.20.1
 	k8s.io/apimachinery => k8s.io/apimachinery v0.20.1
+	k8s.io/cli-runtime => k8s.io/cli-runtime v0.20.1
```
Comment (kpango) [inline go.mod:32]: could you please add replacement to hack/go.mod.default ?

**Review by kpango — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e-deploy.yml (hunk 3 above)
Comment (kpango) [reply .github/workflows/e2e-deploy.yml:350]: see this example 
https://github.com/vdaas/vald/blob/master/Makefile.d/k8s.mk#L266

File: .github/chatops_commands.md [hunk 5]
```diff
@@ -4,3 +4,4 @@
 - :white_check_mark: `/gen-test` - generate test codes
 - :label: `/label` - add labels
 - :rewind: `/rebase` - rebase master
+- :end: :two: :end: `/label actions/e2e-deploy` - run E2E deploy & integration test
```
Comment (kpango) [inline .github/chatops_commands.md:7]: seems good, but how about add another option like /e2e which adds actions/e2e-deploy labels, then executes e2e test by event.

File: .github/workflows/e2e-deploy.yml [hunk 6]
```diff
@@ -38,46 +43,73 @@ jobs:
...
+          fi
+          if curl -s "https://registry.hub.docker.com/v2/repositories/vdaas/vald-lb-gateway/tags/pr-${pr_num}" | jq '.name' | grep -v "null"; then
+            export HELM_EXTRA_OPTIONS="${HELM_EXTRA_OPTIONS} --set gateway.lb.image.tag=pr-${pr_num}"
+          fi
+          if curl -s "https://registry.hub.docker.com/v2/repositories/vdaas/vald-meta-gateway/tags/pr-${pr_num}" | jq '.name' | grep -v "null"; then
+            export HELM_EXTRA_OPTIONS="${HELM_EXTRA_OPTIONS} --set gateway.meta.image.tag=pr-${pr_num}"
+          fi
+          if curl -s "https://registry.hub.docker.com/v2/repositories/vdaas/vald-manager-index/tags/pr-${pr_num}" | jq '.name' | grep -v "null"; then
+            export HELM_EXTRA_OPTIONS="${HELM_EXTRA_OPTIONS} --set manager.index.image.tag=pr-${pr_num}"
+          fi
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:87]: how about using map type and loop like https://github.com/vdaas/vald/pull/814/files#diff-baa3fedacddc444b694c875d076520f333b7239a93ffdb6b87d2cf1e79a48182R46

File: .github/workflows/e2e-deploy.yml [hunk 7]
```diff
@@ -93,58 +125,307 @@ jobs:
...
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:222]: same

File: .github/workflows/e2e-deploy.yml (hunk 1 above)
Comment (rinx) [reply .github/workflows/e2e-deploy.yml:250]: Can I change the value? Is it okay?

File: .github/workflows/e2e-deploy.yml (hunk 3 above)
Comment (rinx) [reply .github/workflows/e2e-deploy.yml:350]: thanks!

File: .github/chatops_commands.md (hunk 5 above)
Comment (rinx) [reply .github/chatops_commands.md:7]: thanks, but adding new chatops command takes too much cost. (e.g. maintain, CI processing speed)
so i don't wanna prepare such a thing.

File: .github/workflows/e2e-deploy.yml (hunk 1 above)
Comment (kpango) [reply .github/workflows/e2e-deploy.yml:250]: I think it's okay, if it works.
make scylla commands is for development we don't need to think so serious.
but it's better to overwrite by make ARGUMENT 

```
SCYLLA_CRD_DIR ?=k8s/external/scylla
k8s/external/scylla/deploy:
        kubectl apply -f $(SCYLLA_CRD_DIR)
```

File: .github/chatops_commands.md (hunk 5 above)
Comment (kpango) [reply .github/chatops_commands.md:7]: I understand just suggestion.
please don't mind

File: .github/workflows/e2e-deploy.yml (hunk 1 above)
Comment (rinx) [reply .github/workflows/e2e-deploy.yml:250]: okay thanks!

File: .github/chatops_commands.md (hunk 5 above)
Comment (rinx) [reply .github/chatops_commands.md:7]: sorry for inconvenience :sob:

File: .github/workflows/e2e-deploy.yml [hunk 8]
```diff
@@ -38,113 +43,372 @@ jobs:
...
+            ["vdaas/vald-agent-ngt"]="agent.image.tag"
+            ["vdaas/vald-discoverer-k8s"]="discoverer.image.tag"
+            ["vdaas/vald-manager-compressor"]="manager.compressor.image.tag"
+            ["vdaas/vald-meta-redis"]="meta.image.tag"
+            ["vdaas/vald-manager-backup-mysql"]="manager.backup.image.tag"
+            ["vdaas/vald-backup-gateway"]="gateway.backup.image.tag"
+            ["vdaas/vald-lb-gateway"]="gateway.lb.image.tag"
+            ["vdaas/vald-meta-gateway"]="gateway.meta.image.tag"
+            ["vdaas/vald-manager-index"]="manager.index.image.tag"
+          )
```
Comment (kpango) [inline .github/workflows/e2e-deploy.yml:72]: nice!

**Review by kpango — DISMISSED** [review]
LGTM

**Comment by rinx** [comment]
Thanks, but currently i'm testing it on Actions. please wait for it. :pray:

**Comment by rinx** [comment]
E2E deploy test (Redis, MySQL) & (Cassandra) failed to execute Remove operations in the recently CI runs. maybe deploying all components is too heavy for the GitHub Actions runners.

**Comment by rinx** [comment]
Updated. 
E2E tests for Vald with Scylla and Vald with MySQL & Redis are failing on StreamUpdate operations.

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**